### PR TITLE
Connector Facebook-Marketing: update insights streams with custom entries for fields, breakdowns and action_breakdowns

### PR DESCRIPTION
**Pull request by vladimir-remar**
## What
In our integration with the Facebook-Marketing connector when we trying to synchronize the API data from [insights](https://developers.facebook.com/docs/marketing-api/insights) all attempts were unsuccessful (see logs on Notes). After experiencing the same type of behavior in Stitch, the solution is to request less fields to Facebook API. For thar reason, it is proposed is to add custom dynamic schemas to reduce the number of fields that are sent in each request. This means, users will be able to choose the fields that are requested and obtained in the response. After seeing the positive results both in Stitch and in a custom Facebook-Marketing connector created by us, we propose the following alternative.

## How
The solution is inspired by the connector of [Google-Analytics-Singer](https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/connectors/source-googleanalytics-singer/source_googleanalytics_singer) that adds "custom_reports" that are configured in the config. With this idea we propose the following:

1. Add to the config an optional field *custom_insights_fields* that contains the fields separated by commas that we want to send in each request to the insights api.
2. In the Streams file, create new classes *(CustomAdsInsights, CustomAdsInsightsAgeAndGender, CustomAdsInsightsCountry, CustomAdsInsightsRegion, CustomAdsInsightsDma, CustomAdsInsightsPlatformAndDevice)* with the prefix "Custom" to differentiate them from the rest and that these classes inherit the behavior of the master class *AdsInsights* and only varying the final resulting schema and fields in these new classes.
3. In the Source file modify the *streams* method so that it is updated if the new *custom_insights_fields* field is configured.

## Notes
I'm adding some logs and some captures.
[error.log](https://github.com/airbytehq/airbyte/files/6849623/error.log)
Screenshot from 2021-07-20 17-45-24
Screenshot from 2021-07-20 17-46-15

I acknowledge this first commit is missing the Documentation updates and the docker version bump.

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [ ] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by sherifnada** [comment]
@vladimir-remar would this be supplanted by the solution in #2227 ?

**Comment by vladimir-remar** [comment]
@sherifnada Yes, but, depending on the implementation, the Facebook-Marketing connector will need some additional changes as the request to the insights API has to explicitly include the limited number of fields.
Since we do not know when solution to #2227 will be implemented, does it make sense to merge the solution I propose in the PR? The only way to synchronize the Facebook-Marketing data with our destination, as of now, is to implement a custom Facebook connector.

**Comment by vladimir-remar** [comment]
@sherifnada any thoughts or suggestions?

**Comment by marcosmarxm** [comment]
@keu is possible to review this contribution?

**Comment by vladimir-remar** [comment]
Hi again.
After this time it makes sense to think of an approach more similar to the one proposed in this [PR](https://github.com/airbytehq/airbyte/pull/5302), as I said the only way we can synchronize the data from facebook is using the custom streams (at least for now).
I also take into account the https://github.com/airbytehq/airbyte/issues/2227 but I don't know the status.

**Comment by keu** [comment]
@vladimir-remar Insights is already pretty custom streams. As I remember after my last refactor it is pretty easy to make it fully custom. The only reason why I didn't is that it breaks backward compatibility.

So the end result will replace all insights streams with a list of configuration objects in the config.
```json
{
"insights":
 [
   {
    "name": "text",
    "fields": [],
    "breakdowns": [],
    "action_breakdowns": [],
   }
 ]
}
```

**Comment by vladimir-remar** [comment]
> @vladimir-remar Insights is already pretty custom streams. As I remember after my last refactor it should be easy to make it fully custom. The only reason why I didn't is that it breaks backward compatibility.
> 
> You just need to move all hardcoded attributes to config and build instances of the `InsightsStream` in `streams` method.
> So the end result will replace all insights streams with a list of configuration objects in the config.
> 
> ```json
> {
> "insights":
>  [
>    {
>     "name": "text",
>     "fields": [],
>     "breakdowns": [],
>     "action_breakdowns": [],
>    }
>  ]
> }
> ```

Ok, for each Insights stream we will have a different configuration and if some insights were not in the json object we would have to define default values for the streams that are not included in this object, right?

**Comment by vladimir-remar** [comment]
@keu Hello, I did some changes in order to improve the new approach, can you take a look?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py [hunk 1]
```diff
@@ -76,6 +89,7 @@ class Config:
         minimum=1,
         maximum=30,
     )
+    custom_insights: Optional[Json] = Field(description="A json objet with custom insights")
```
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:92]: I think you need to declare a new model here - `InsightConfig` and have a list of it here

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py [hunk 2]
```diff
@@ -21,14 +21,20 @@
...
+from airbyte_cdk.models import (
+    ConnectorSpecification,
+    DestinationSyncMode,
+    AirbyteCatalog
+)
 from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.logger import AirbyteLogger
 from airbyte_cdk.sources.streams import Stream
-from pydantic import BaseModel, Field
+from pydantic import BaseModel, Field, Json
```
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:37]: not needed, see comments below

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py [hunk 3]
```diff
@@ -41,6 +47,13 @@
     AdsInsightsPlatformAndDevice,
     AdsInsightsRegion,
     Campaigns,
+    CustomAdsInsights,
```
Comment (keu) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:50]: don't think you actually need these anymore

**Review by keu — CHANGES_REQUESTED** [review]
great job in general, just a few comments

**Comment by marcosmarxm** [comment]
@vladimir-remar awesome work! thanks so much for this contribution. Small change, could you solve the conflict in source.py?

@keu can you final review this contribution?

**Comment by vladimir-remar** [comment]
> @vladimir-remar awesome work! thanks so much for this contribution. Small change, could you solve the conflict in source.py?
> 
@marcosmarxm done!

**Review by keu — APPROVED** [review]
lgtm

**Comment by keu** [comment]
@vladimir-remar don't forget to `gradle format` and tests

**Comment by vladimir-remar** [comment]
> @vladimir-remar don't forget to run `./gradlew format` and tests

done

**Comment by vladimir-remar** [comment]
@keu You want me to dump the versions?

**Comment by keu** [comment]
@vladimir-remar Thanks.
No need to bump, I will do it myself, just let me ask you to add CHANGELOG to the `facebook-marketing.md` file.

**Comment by vladimir-remar** [comment]
> @vladimir-remar Thanks. No need to bump, I will do it myself, just please add CHANGELOG to the `facebook-marketing.md` file.

@keu done!

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py [hunk 4]
```diff
@@ -126,3 +148,28 @@ def spec(self, *args, **kwargs) -> ConnectorSpecification:
...
+        default insights streams else returns streams
+
+        """
+        if not insights:
+            return streams
+
+        insights_custom_streams = list()
+
+        for insight in insights:
+            args["name"] = insight.name
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:163]: is this mutating the same config over and over, so in the end all the streams have the same config? can we copy the args instead? e.g: 

```
arg_copy = copy.deepcopy(args)
arg_copy["fields"] = ...
```

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py [hunk 5]
```diff
@@ -126,3 +148,28 @@ def spec(self, *args, **kwargs) -> ConnectorSpecification:
                 ),
             ),
         )
+
+    def _update_insights_streams(self, insights, args, streams) -> List[Type[Stream]]:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:152]: I suggest the following approach for these custom streams: 

1. The "standard" streams offered by the connector are always available (if a user doesn't want them they can always just deselect them)
2. If the user inputs any custom streams, they are named as `custom_<user-input-name>` and appended to the list of streams

This way it is very very obvious to the user what is happening. This is especially important as the connector's config is updated over time e.g: a user might call a stream `ads_insights` today, then remove it next week, and now the data represented in that stream is mixed with the "standard" ads insight stream and the "custom" insight stream. Making the suggestion above to make such situations much less likely.

**Review by sherifnada — COMMENTED** [review]
@vladimir-remar thanks for the contribution! I certainly understand the value of having something like this that works well. Some points I'd like to make: 

I'm wary of introducing a completely custom insights stream because the Facebook API is very volatile. Specifically: it is difficult to predict which inputs of fields/breakdowns/action_breakdowns are going to cause the insights sync to succeed or fail. For all of our current insights streams, we've done meticulous testing to make sure those requests succeed consistently. So the issue I'm worried about is a user might see this feature, try to use it, get a failure, and then assume the FB connector is unreliable and file support requests with Airbyte when really the issue is from the FB side. The error messages provided by the FB API tend to be pretty opaque too and unhelpful, e.g: `Job Failed` without any extra information. 

This is all to say that, I think all of this should be made very very clear in the connector's documentation (`docs/integrations/sources/facebook-marketing.md`), that this is a "use with caution" or "experimental" feature. We should add a section about these custom insights streams in the docs describing all the above caveats. 

Additionally, it would be great to have a way to tell the user ahead of time if their particular selection will work or not. For example, right now if they choose an incorrect field name, they will only know that problem at sync time not connector setup time. 

WDYT?

**Comment by vladimir-remar** [comment]
To @keu thank you for help me in the spec limitation, was very helpful.
To @sherifnada I added some changes.
- A way to verify the insights before the sync, in the check part.
- An update to the documentation, warning the use for users.

**Review by sherifnada — COMMENTED** [review]
@vladimir-remar thanks for adding the validations! just a couple of comments and I think we are good to go

**Review by sherifnada — APPROVED** [review]
LGTM, thanks @vladimir-remar !

@keu could you assign to someone on the team for publishing?

**Comment by zestyping** [comment]
This is awesome!! Thank you so much!

**Pull request closed without merging**